zilliqa/src/consensus.rs

```diff
@@ -999,83 +1032,148 @@ impl Consensus {
         Ok(None)
     }
 
-    fn propose_new_block_at(
+    /// Finalise the early Proposal.
+    /// This should only run after majority QC is available.
+    /// It applies the rewards and produces the final Proposal.
+    fn finish_early_proposal_at(
         &self,
         state: &mut State,
-        transaction_pool: &mut TransactionPool,
-        votes: &mut BTreeMap<Hash, BlockVotes>,
-    ) -> Result<Option<(Block, Vec<VerifiedTransaction>)>> {
-        let num = self
-            .block_store
-            .get_highest_block_number()
-            .unwrap()
-            .unwrap();
-        let block = self.get_block_by_number(num).unwrap().unwrap();
+        votes: &BTreeMap<Hash, BlockVotes>,
+        proposal: Block,
+    ) -> Result<Option<Block>> {
+        // Retrieve parent block data
+        let parent_block = self
+            .get_block(&proposal.parent_hash())?
+            .context("missing parent block")?;
+        let parent_block_hash = parent_block.hash();
+        let parent_block_view = parent_block.view();
+
+        // Should have supermajority by now.
+        // Retrieve the committee - for rewards
+        let committee = state.get_stakers_at_block_raw(&parent_block)?;
+        let (signatures, cosigned, _, _) = votes
```

Comment:
We shouldn't propose a block without supermajority QC/AggQC, it would only spam the other valudators and they would reject it anyways. If we reach supermajority before we time out, we propose a block. If we don't reach supermajority and time out, we advance to the next view and don't care about belated votes that we should have received in the previous view.

Reply:
Alright. In that case I'll add a check to fail if the code ever gets called w/o a supermajority.